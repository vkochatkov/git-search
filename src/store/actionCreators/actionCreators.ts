import { Action, AnyAction, Dispatch } from 'redux';

import axios from 'axios';
import {
  CLEAR_USERS,
  GET_REPOS,
  GET_USER,
  SEARCH_USERS,
  SET_LOADING,
} from '../actionTypes';
import { ReposProps, User, Users } from '../../interfaces/interface';
import { CLIENT_ID, CLIENT_SECRET } from '../../constants/contsants';

const withCreds = (url: string) =>
  `${url}client_id=${CLIENT_ID}&client_secret=${CLIENT_SECRET}`;

export const setLoading = (): Action => ({ type: SET_LOADING });

export const clearUsers = (): Action => ({ type: CLEAR_USERS });

export const setSearchQuery = (users: Users) => ({
  type: SEARCH_USERS,
  payload: users,
});

export const setUser = (user: User) => ({ type: GET_USER, payload: user });

export const setRepos = (repos: ReposProps) => ({
  type: GET_REPOS,
  payload: repos,
});

export const setError = () => ({});

export const search = (value: string) => {
  return async (dispatch: Dispatch) => {
    dispatch(setLoading());
    const response = await axios.get(
      withCreds(`https://api.github.com/search/users?q=${value}&`),
    );

    dispatch(setSearchQuery(response.data.items));
  };
};

export const getUser = (name: string) => {
  return async (dispatch: Dispatch) => {
    try {
      dispatch(setLoading());

      const response = await axios.get(
        withCreds(`https://api.github.com/users/${name}?`),
      );

      dispatch(setUser(response.data));
    } catch (e) {
      dispatch(setError());
    }
  };
};

export const getRepos = (name: string) => {
  return async (dispatch: Dispatch) => {
    dispatch(setLoading());

    const response = await axios.get(
      withCreds(`https://api.github.com/users/${name}/repos?per_page=10&`),
    );

    dispatch(setRepos(response.data));
  };
};
