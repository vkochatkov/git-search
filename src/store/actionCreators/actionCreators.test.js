import configureMockStore from 'redux-mock-store';
import { setLoading, setRepos, search } from './actionCreators';
import thunk from 'redux-thunk';
import nock from 'nock';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('async actions', () => {
  let store;
  let fetchedData = [
    {
      avatar_url: 'https://avatars1.githubusercontent.com/u/5108906?v=4',
      events_url: 'https://api.github.com/users/vitaly-t/events{/privacy}',
      followers_url: 'https://api.github.com/users/vitaly-t/followers',
      following_url:
        'https://api.github.com/users/vitaly-t/following{/other_user}',
      gists_url: 'https://api.github.com/users/vitaly-t/gists{/gist_id}',
      gravatar_id: '',
      html_url: 'https://github.com/vitaly-t',
      id: 5108906,
      login: 'vitaly-t',
      node_id: 'MDQ6VXNlcjUxMDg5MDY=',
      organizations_url: 'https://api.github.com/users/vitaly-t/orgs',
      received_events_url:
        'https://api.github.com/users/vitaly-t/received_events',
      repos_url: 'https://api.github.com/users/vitaly-t/repos',
      score: 1,
      site_admin: false,
      starred_url:
        'https://api.github.com/users/vitaly-t/starred{/owner}{/repo}',
      subscriptions_url: 'https://api.github.com/users/vitaly-t/subscriptions',
      type: 'User',
      url: 'https://api.github.com/users/vitaly-t',
    },
  ];

  beforeEach(() => {
    store = mockStore({});
  });
  afterEach(() => {
    nock.cleanAll();
  });

  it('creates FETCH_TODO_SUCCESS when fetching to-dos has been done', () => {
    nock('https://localhost:8000').get('/api/users').reply(200, fetchedData);

    return store.dispatch(search('kochatkov')).then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
  });
});
