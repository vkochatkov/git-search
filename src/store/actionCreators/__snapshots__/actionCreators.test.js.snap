// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`async actions creates FETCH_TODO_SUCCESS when fetching to-dos has been done 1`] = `
Array [
  Object {
    "type": "SET_LOADING",
  },
  Object {
    "payload": Array [
      Object {
        "avatar_url": "https://avatars3.githubusercontent.com/u/61251470?v=4",
        "events_url": "https://api.github.com/users/kochatkov/events{/privacy}",
        "followers_url": "https://api.github.com/users/kochatkov/followers",
        "following_url": "https://api.github.com/users/kochatkov/following{/other_user}",
        "gists_url": "https://api.github.com/users/kochatkov/gists{/gist_id}",
        "gravatar_id": "",
        "html_url": "https://github.com/kochatkov",
        "id": 61251470,
        "login": "kochatkov",
        "node_id": "MDQ6VXNlcjYxMjUxNDcw",
        "organizations_url": "https://api.github.com/users/kochatkov/orgs",
        "received_events_url": "https://api.github.com/users/kochatkov/received_events",
        "repos_url": "https://api.github.com/users/kochatkov/repos",
        "score": 1,
        "site_admin": false,
        "starred_url": "https://api.github.com/users/kochatkov/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/kochatkov/subscriptions",
        "type": "User",
        "url": "https://api.github.com/users/kochatkov",
      },
    ],
    "type": "SEARCH_USERS",
  },
]
`;
